src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs

```diff
@@ -381,5 +430,38 @@ public void BytesRead(int count)
         {
             Interlocked.Add(ref _readTimingBytesRead, count);
         }
+
+        public void StartTimingWrite(int size)
+        {
+            lock (_writeTimingLock)
+            {
+                var minResponseDataRate = _frame.MinResponseDataRate;
+
+                if (minResponseDataRate != null)
+                {
+                    var timeoutTicks = Math.Max(
+                        minResponseDataRate.GracePeriod.Ticks,
+                        TimeSpan.FromSeconds(size / minResponseDataRate.BytesPerSecond).Ticks);
+
+                    if (_writeTimingTimeouts.Count == 0)
+                    {
+                        _writeTimingLastTimeoutTimestamp = _lastTimestamp;
+                    }
+
+                    // Add Heartbeat.Interval since this can be called right before the next heartbeat.
+                    _writeTimingLastTimeoutTimestamp += timeoutTicks + Heartbeat.Interval.Ticks;
+
+                    _writeTimingTimeouts.Enqueue(_writeTimingLastTimeoutTimestamp);
```

Comment:
Sounds good. It was my initial thought, but I figured it would be easier to remove the queue if I got that feedback than add it later 😁